accepted/000070-cluster_awareness.md

```diff
@@ -0,0 +1,272 @@
+
+Uyuni/SUSE Manager will store all entered cluster details data into the database and after successful provisioning of the management node, the cluster overview will be shown.
+
+
+### Cluster overview
+
+Under “Cluster overview”, all clusters known to Uyuni/SUSE Manager will be displayed. When a cluster is selected, a new page listing all the details of the cluster will be shown, containing:
+
+- List of all nodes of the cluster - by invoking the cluster provider manager (via Salt) on the Management Node:
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.list_nodes <cluster name>
+  ```
+    - For CaaSP: `salt <mgmt node> skuba.list_nodes <cluster name>`
+    - For SES: `salt <mgmt node> ceph.list_nodes <cluster name>`
+    - For SLE-HA/SHAP: `salt <mgmt node> crm.list_nodes <cluster name>`
+
+- Optional information retrieved via Salt grains (populated by the product):
+    - Link to the product dashboard (e.g. SES)
+    - Credentials download (e.g. download `kubeconfig` from CaaSP)
+- Optional information regarding the health of each node of the cluster (if supported by the cluster).
+
+By enumerating all the functions of the Salt module corresponding to the cluster provider manager, Uyuni/SUSE Manager will offer to run all the actions that the cluster provider manager exposes.
+A minimum list of actions that must be exposed by the cluster provider manager is:
+
+* Adding a node: Uyuni/SUSE Manager will invoke the cluster provider manager on the management node of the cluster and will pass any optional parameter (requested to the user) (e.g. username, IP of the system, type of the node).
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.add_node <cluster name> <optional params>
+  ```
+
+  All the parameters of the formula have to be specified by the cluster provider manager's types.
+
+  In a future version, Uyuni/SUSE Manager may offer the possibility to present to the user a drop-down selection of systems registered to Uyuni/SUSE Manager and not part of any cluster.
+
+* Removing a node: Uyuni/SUSE Manager will invoke the cluster provider manager on the management node of the cluster and will pass any optional parameter (requested to the user) to the call using Salt grains or by using a Salt formula.
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.remove_node <cluster name> <optional params>
+  ```
+
+* Node reboot: Uyuni/SUSE Manager will invoke the cluster provider manager on the management node of the cluster and will pass any optional parameter (requested to the user) to the call using Salt grains or by using a Salt formula.
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.reboot_node <cluster name> <optional params>
```

Comment:
I get the reboot part (which basically just schedules a reboot for later because the admin has done something that requires a reboot). This is already violating the control flow though, because technically there should be no reason for having to do that reboot other than when the cluster manager does it on its own.
Actually mandating the _update_ of a single node is a clear violation of control flow. This should _always_ be triggered by the cluster manager once we've asked for a cluster update through the API.

Reply:
> Do we (also) plan for an endpoint to upgrade a single node in the cluster, including any reboots if required?

I think offering it would result in incoherent situations: a cluster node is updated and other nodes are not.

Reply:
You are right in saying such a feature needs to be used carefully, as it opens up more failure cases (wrong ordering, incomplete upgrades, likely more).

OTOH, I remember several things mentioned during the last Winter Summit that made me leave the event with the feeling that supporting individual node upgrades was not really something we could avoid.

In particular, IIRC:
 - a whole-cluster upgrade procedure is not really sufficient and might not even be wanted, as in some cases manual steps between one node and another might be needed (at least for HA)
 - the exact ordering of the nodes to upgrade might also depend on particular customer needs (at least for SES). It was mentioned that some users appear to be very picky about what happens to their machines and tend to be extra conservative, really demanding control of every single step (for example, draining is not always wanted, but sometimes it is)
 - there are scenarios in which users want to do extra manual maintainance on a node while it is marked "down" - and those might even be completely outside of the scope and control of our products/projects (at least for HA). "Sysadmin paranoia" was also mentioned in cases where HA is used as infrastructure for SAP, understandably
 - there is currently no functionality for whole-cluster upgrade in HA - upgrades are always carried out manually or semi-automatically on a node-by-node basis. I checked the documentation linked from this PR and - unless I am missing something - I indeed cannot find references to existing functionality at this point, and undoubtedly this is not something we want to implement in Uyuni
 - someone observed that a single node upgrade feature would be useful to recover from an incomplete situation caused by manual intervention

Please accept my apologies if I am misremembering something, or my notes were not correct, or my knowledge is not up-to-date at this point, but please consider double checking the above as I fear we might be missing some important aspect.

Reply:
I completely understand your point and it makes sense to me as well.

To take the correct decision, I gathered some data:

|   | CaaSP | SES | SHAP |
| ------------- | ------------- | ------------- |
| Individual node upgrade | skuba-update  | [N/A in SES 7](https://chat.suse.de/channel/ceph?msg=JpA8rqEYjNCQSSD5F) | Enable maintenance mode; `zypper dup`; Disable maintenance mode |
| All nodes upgrade  | skuba and with a predefined order (control-planes first, workers after)  | [N/A in SES 7](https://chat.suse.de/channel/ceph?msg=JpA8rqEYjNCQSSD5F) | | needs to be orchestrated - no predefined order |

Regarding the "All nodes upgrade" and the order, as well as what to do in case something goes wrong, I would delegate the strategy and the recovery to the cluster provider manager. From the Uyuni/SUSE Manager POV, we should just issue a cluster upgrade and then check for the result (good or bad).

About the Individual node upgrade: as per the reboot, it is an action that is in the scope of a single node. I see where your argument comes from and I agree. To me, it might make sense to move "node reboot" and "node upgrade" to the already existing single system detail page, but still, delegate the operations to the cluster provider manager.
This will open up the possibility of doing the upgrade on a node-by-node basis, as well as offering the possibility of troubleshooting you were correctly mentioning.

About mixed version situation that one partial upgrade might result in: SHAP explictly warns user not to mix between different product versions (https://documentation.suse.com/sle-ha/15-SP1/single-html/SLE-HA-guide/#sec-ha-migration-upgrade) and, in case of rolling upgrades, to "complete the whole cluster upgrade in one week" (https://documentation.suse.com/sle-ha/15-SP1/single-html/SLE-HA-guide/#sec-ha-migration-upgrade-rolling).
We should strive from completely avoiding ending up in this situation by avoiding single-node upgrade, especially for SHAP.

And here you see the conundrum of the upgrade situation. IMHO my proposal would be: move individual node actions (reboot and upgrade) to the future developments section. Luckily, in the current RFC only the reboot action has been written, so the proposal boils down to only move the single node reboot action.

Do you have a different proposal?

Reply:
> Regarding the "All nodes upgrade" and the order, as well as what to do in case something goes wrong, I would delegate the strategy and the recovery to the cluster provider manager. From the Uyuni/SUSE Manager POV, we should just issue a cluster upgrade and then check for the result (good or bad).

Agreed.

> About the Individual node upgrade: as per the reboot, it is an action that is in the scope of a single node. I see where your argument comes from and I agree. To me, it might make sense to move "node reboot" and "node upgrade" to the already existing single system detail page, but still, delegate the operations to the cluster provider manager.
> This will open up the possibility of doing the upgrade on a node-by-node basis, as well as offering the possibility of troubleshooting you were correctly mentioning.

Agreed.

> About mixed version situation that one partial upgrade might result in: SHAP explictly warns user not to mix between different product versions (https://documentation.suse.com/sle-ha/15-SP1/single-html/SLE-HA-guide/#sec-ha-migration-upgrade) and, in case of rolling upgrades, to "complete the whole cluster upgrade in one week" (https://documentation.suse.com/sle-ha/15-SP1/single-html/SLE-HA-guide/#sec-ha-migration-upgrade-rolling).
> We should strive from completely avoiding ending up in this situation by avoiding single-node upgrade, especially for SHAP.

I think we should limit the scope of this RFC to regular updates only, ie. not major version upgrades (eg. SES 6 to 7). Those might always need one-shot special procedures that might not make sense to automate at all. Basically, the fact they are one-shot efforts means that automating does not add a lot of value.

FMPOV what Manager/Uyuni should focus on is the normal day-to-day operations, especially OS and security updates. Those get much more value from automation, because they are recurring frequently.

> And here you see the conundrum of the upgrade situation. IMHO my proposal would be: move individual node actions (reboot and upgrade) to the future developments section. Luckily, in the current RFC only the reboot action has been written, so the proposal boils down to only move the single node reboot action.
> 
> Do you have a different proposal?

My personal thinking:
 - automating (regular, not major) updates is the single most important feature of this whole initiative, and actually (if the concept is seen as general, including single machines or containers) of the whole SUSE Manager product and Uyuni project
 - HA does not have a whole-cluster update capability, so not having individual node upgrades in this RFC means substantially excluding HA
 - SES 7 might or might not have a whole-cluster update capability, still customers often demand individual node updates. So not having individual node upgrades in this RFC means to substantially lower the value proposition for that product
 - because of the above points, I think individual node updates should positively be part of the MVP
 - conversely, we could in principle live without a whole-cluster option, at least in a first iteration. CaaSP users might or might not miss it, I have not heard from the field - but arguably they would still the main benefit of Manager/Uyuni with an individual node update option alone

That said, I am not authoritative on priorities as you know, so I will leave it up to you and relevant folks to figure out the actual best course of action.

HTH!

Reply:
>     * automating (regular, not major) updates is the single most important feature of this whole initiative, and actually (if the concept is seen as general, including single machines or containers) of the whole SUSE Manager product and Uyuni project
> 
>     * HA does not have a whole-cluster update capability, so not having individual node upgrades in this RFC means substantially excluding HA
> 
>     * SES 7 might or might not have a whole-cluster update capability, still customers often demand individual node updates. So not having individual node upgrades in this RFC means to substantially lower the value proposition for that product
> 
>     * because of the above points, I think individual node updates should positively be part of the MVP

Unless I'm missing something, whole-cluster update implies single-node update, at least for the cluster manager.

The main difference is whether **Uyuni** will be able to do whole-cluster or single-node updates.

If HA or SES do not have a "whole-cluster update" concept or mechanism, it must be implemented by the cluster manager. All Uyuni should know is it can invoke a whole-cluster update and the cluster manager will perform its magic behind the scenes. Whether that magic is invoking an actual whole-cluster update, or it is orchestrating single-node updates **inside** the cluster manager (i. e. the details on how that operation is implemented do not leak to Uyuni), we don't care.

Does that make sense?

Reply:
>     * HA does not have a whole-cluster update capability, so not having individual node upgrades in this RFC means substantially excluding HA

A cluster upgrade for SHAP would be translated (into the cluster provider manager) into:

- enables maintenance mode for all nodes of the cluster
- `zypper dup` for all nodes of the cluster
- disable maintenance mode for all nodes of the cluster 

>     * SES 7 might or might not have a whole-cluster update capability, still customers often demand individual node updates. So not having individual node upgrades in this RFC means to substantially lower the value proposition for that product

As per the table above, right now we do not have enough data about SES 7. The value of the RFC reflects the current status of the research and the information available.

Reply:
> If HA or SES do not have a "whole-cluster update" concept or mechanism, it must be implemented by the cluster manager. All Uyuni should know is it can invoke a whole-cluster update and the cluster manager will perform its magic behind the scenes. Whether that magic is invoking an actual whole-cluster update, or it is orchestrating single-node updates **inside** the cluster manager (i. e. the details on how that operation is implemented do not leak to Uyuni), we don't care.

Yes, what you wrote aligns with the current status of this RFC, except for the individual ode reboot action that I was questioning whether it should be part of the MVP or now because of https://github.com/uyuni-project/uyuni-rfc/pull/36/files#r395110633